File: Hard/847. Shortest Path Visiting All Nodes/847. Shortest Path Visiting All Nodes.ts
function shortestPathLength(graph: number[][]): number {
    const n: number = graph.length;
    const targetMask: number = (1 << n) - 1;
    const queue: [number, number, number][] = [];
    const visited: boolean[][] = new Array(n)
        .fill(null)
        .map(() => new Array(1 << n).fill(false)); // DP table to store visited states
    for (let i = 0; i < n; i++) {
        queue.push([i, 1 << i, 0]); // Initialize the queue with starting nodes
        visited[i][1 << i] = true;
    }
    let steps: number = 0;

    while (queue.length > 0) {
        const levelSize: number = queue.length;
        for (let i = 0; i < levelSize; i++) {
            const [node, mask, cost]: [number, number, number] = queue.shift()!;
            if (mask === targetMask) {
                return cost;
            }
            for (const neighbor of graph[node]) {
                const newMask: number = mask | (1 << neighbor);
                if (!visited[neighbor][newMask]) {
                    visited[neighbor][newMask] = true;
                    queue.push([neighbor, newMask, cost + 1]);
                }
            }
        }
        steps++;
    }

    return -1; // Return -1 if no path exists (this should not occur for connected graphs).
}

File: Hard/847. Shortest Path Visiting All Nodes/847. Shortest Path Visiting All Nodes.java
import java.util.ArrayDeque;
import java.util.Queue;

class Solution {
    public int shortestPathLength(int[][] graph) {
        int n = graph.length;
        int targetMask = (1 << n) - 1;
        Queue<int[]> queue = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][1 << n]; // DP table to store visited states

        for (int i = 0; i < n; i++) {
            queue.offer(new int[] { i, 1 << i, 0 }); // Initialize the queue with starting nodes
            visited[i][1 << i] = true;
        }

        int steps = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                int[] nodeInfo = queue.poll();
                int node = nodeInfo[0];
                int mask = nodeInfo[1];
                int cost = nodeInfo[2];

                if (mask == targetMask) {
                    return cost;
                }

                for (int neighbor : graph[node]) {
                    int newMask = mask | (1 << neighbor);
                    if (!visited[neighbor][newMask]) {
                        visited[neighbor][newMask] = true;
                        queue.offer(new int[] { neighbor, newMask, cost + 1 });
                    }
                }
            }
            steps++;
        }

        return -1; // Return -1 if no path exists (this should not occur for connected graphs).
    }
}
